### ISSUE TITLE
something wrong in the S2A model

### ISSUE THREAD

**Issue by yangdongchao** [collabora/WhisperSpeech#19]
hi,
   I carefully read your S2A model code, i found that you try to model 2 codebook, when you transfer the codebook tokens into vector using nn.Embedding, you assume the first codebook's token and the second codebook's token using the same nn.Embedding, which means that you assume that the two codebook's token enjoy the same  data distributions. I suggest you use different nn.Embedding to deal with different codebook's token.
 Maybe my suggestion is wrong, I am willing to discuss with you.

**Comment by jpc** [comment]
Hey, you are right – this looked wrong to me as well but it seems like it does not matter in practice. I suspect the positional embeddings are making it easy to learn to treat every other token differently (it's also visible in attention scores – they attend "in pairs").

I tried using separate codebooks at the beginning but the validation cross-entropy loss was basically the same and it complicated the autoregressive sampling code (every other step you have an odd number of tokens and you have to work around that).

**Comment by yangdongchao** [comment]
> Hey, you are right – this looked wrong to me as well but it seems like it does not matter in practice. I suspect the positional embeddings are making it easy to learn to treat every other token differently (it's also visible in attention scores – they attend "in pairs").
> 
> I tried using separate codebooks at the beginning but the validation cross-entropy loss was basically the same and it complicated the autoregressive sampling code (every other step you have an odd number of tokens and you have to work around that).

Yes, using positional embedding seems can inform the network. But I still recommend you to use different nn.Embeding for different codebooks. In my early experiments, it has been validated as a better choice.

**Comment by jpc** [comment]
Hey, thanks for the feedback. I think I will look into integrating the SoundStorm paper and then I would have to use separate embeddings since all then all tokens from the same timestep are used at the same position.

**Comment by jpc** [comment]
The current model I am training is based on the multi-head delayed autoregressive architecture of MusicGen so it has separate codebooks for each quantizer layer.
